state/storage.go

```diff
@@ -879,11 +873,16 @@ func (s *Storage) visibleEventNIDsWithData(joinTimingsAtFromByRoomID map[string]
 		if isJoined {
 			result = append(result, [2]int64{startIndex, toIncl})
 		}
-		return result
+		if len(result) == 0 {
+			return [2]int64{}
+		}
+		// we only care about the LAST nid range, otherwise we can end up with gaps being returned in the
+		// timeline. See https://github.com/matrix-org/sliding-sync/issues/365
```

Comment:
I think this wording is going to confuse my future self. (Ditto for the other comment citing this issue). It makes it sound like "gaps in the timeline" is fundamentally bad---but we _do_ want there to be events not sent to users in certain situations (e.g. `joined` history vis).

"Gaps in the timeline" means presumably means(?): the proxy returns a timeline [A, ...., B, C, ... D] such that
1. B is not a direct ancestor of C,
2. there is some event between B and C that the user has permission to see, and
3. there is nothing in the final sync response which tells clients about points (1) and (2)?

Maybe we can say "otherwise we can end up with _incorrect_ gaps in the timeline"?

Reply:
I dunno, for me this seems clear. A "gap" fundamentally is in between two things. In this case, there is a gap between two timeline events e.g:
```
[A,B,C,D,E] true timeline
[A,B,D,E]   calculated timeline with a gap
```

Hence, "gaps in the timeline" _are_ fundamentally bad, as it means we have omitted an event.

Technically, we could be pedantic and say that a "gap" is also present when you have:
```
[A,B,Leave, .... , Join, D, E]
```

and that this is a "correct" gap (assuming joined history visibility), but this has a ton of caveats and ultimately isn't something the proxy should ever do anymore as of this PR. The moment we start differentiating between correct (which won't be producible by the proxy) and incorrect gaps, it just adds more ultimately unnecessary lexicon imo.

This is further supported by how end-users describe it in bug reports:
 - "I have gap in my timeline"
 - "There are gaps in my timeline"
 - "Sent a message from my matrix.org account. Got notification on EIX, opened it from there. When the timeline loaded, there is a gap in the timeline"

Let's use the same term to mean the same thing, a bad thing.

Reply:
it also doesn't help that we speak of "gappy" v2 syncs etc.